### DISCUSSION TITLE
Need a proper API for Codewars clans (feature request)

### DISCUSSION THREAD

**Discussion by khanhtranngoccva** [codewars/codewars.com#2710] [Ideas]
There is no proper API to get a clan member's user information from the clan name yet in the v1 API. I and another folk at #100devs independently created live leaderboards for the #100devs clan, but since this kind of API does not exist, we both had to procure a list of following from a fellow user's social tab (and capture them using browser console JavaScript) and make thousands of HTTP requests to your API (not good).

I'd appreciate a new API that accepts a clan name and return the information of every user in it.

**Comment by kazk** [comment]
This is a duplicate of #760. There are more open feature requests related to clans, and we haven't done anything.

The main problem is, **there's no such thing as a clan**. It's just a text field that simply [triggers users with the same value to follow each other on change](https://docs.codewars.com/community/following#clans).

That said, this one doesn't require modeling anything, so I'm considering to hack something together quickly to stop the leaderboard DDoS :D

`GET /api/v1/clans/{name}/members` that returns a paginated list of members with just enough information for a simple leaderboard. Not a _proper_ API, but something good enough for your use case.

**Comment by khanhtranngoccva** [comment]
~~By the way, can you give me the URL for this community name? I tried encodeURIComponent() but it doesn't work properly well.~~

~~Full name: "#100Devs - leonnoel.com/twitch"
Link: https://www.codewars.com/api/v1/clans/%23100Devs%20-%20leonnoel.com%2Ftwitch~~

Oh! You haven't deployed the API yet! I'm rooting for it so that I can add that leaderboard back to my portfolio, thank you!

**Comment by kazk** [comment]
Just deployed it. It's `GET /api/v1/clans/:name/members` that responds with paginated JSON that looks like

```typescript
type Response = {
  totalPages: number,
  totalItems: number,
  data: Member[], // ordered by descending honor points
};
type Member = {
  id: string,
  username: string,
  honor: number, // honor points
  score: number, // rank score
  rank: number, // -8 to -1 are kyu, 1+ are dan
  languages: {score: number, rank: number}[],
};
```
Specify the page with `?page=2`. Page size is `100`.

Example: https://www.codewars.com/api/v1/clans/Codewars/members

I'll see how this performs and make any adjustments.

---

I didn't realize you guys had a weird clan name like that :man_facepalming: That won't work because of slash and period.

https://www.codewars.com/api/v1/clans/100devs/members is unrelated?

**Reply by Madjosz** [reply]
Edit: I just see that you even mention that in the post. Should have read until the end...

When the clan name contains a `"."`, you are getting the Codewars 404 error page as response opposing to a normal 404 JSON response, even when the clan exists (compare e.g. `modin.us`, `Г.Р.Я.З.Ь`). In the above mentioned `#100Devs` https://www.codewars.com/api/v1/clans/%23100Devs%20-%20leonnoel.com%2Ftwitch/members because of the leading `#` the browser also then interprets it as anchor and cuts off everything after `clans/`.

**Reply by kazk** [reply]
Yeah, name must be URI encoded. `.` is invalid here because of how Rails routing works by default. It's getting global 404 because it doesn't match any of the defined routes.

**Reply by kazk** [reply]
Period is now allowed.

**Reply by khanhtranngoccva** [reply]
Sorry @kazk for the inconvenience! Our community has a weird name 😂

**Reply by khanhtranngoccva** [reply]
Hey there! I'm still having a problem with CORS. Has your API's CORS been enabled yet?

**Reply by khanhtranngoccva** [reply]
Also, we'd love to have the JavaScript score enabled - after all, we are a JS developer community, and it seems that browsing the page from 17-29 does not work - the request freezes and times out on NodeJS and browser JS as well.

**Reply by kazk** [reply]
CORS is enabled for this. But it's failing with your clan name. `modin.us` works, so it's probably the slash.

> it seems that browsing the page from 17-29 does not work

Confirmed with `curl`. I'll take a look.

**Reply by kazk** [reply]
The timeout on pages 17+ should be fixed.

**Comment by kazk** [comment]
@khanhtranngoccva Like I wrote above, CORS has been enabled from the beginning, but it's having trouble because of the slash in the name (it's configured with https://github.com/cyu/rack-cors). I don't have the time to look into this further.

I can change the clan name to something else if you'd like.

**Reply by khanhtranngoccva** [reply]
No thanks, Kazk! I think I can use it now. Will inform you later if something goes wrong!
